Remote run constraints support

keymone opened the pull request:
constraints for #1087 

this supports EQUALS, LIKE, UNLIKE, MAX_PER and UNIQUE

in CLI this adds repeatable adrgument to remote-run: `--cons <attr>,<OP>[,<val>]`, in paasta_remote_run, `--constraints-json`

all constraint-related code is in paasta_tools/frameworks/constraints.py, in native scheduler we have `self.constraints` read from configs or cli (cli has precedence), `self.constraint_state` and `self.constraint_state_lock` for stateful constraints (only max_per for now), state is swapped every time call to `driver.acceptOffers` succeeds. state is locked around processing each offer and in statusUpdate when processing dead tasks.

keymone commented:
what do we set some limit on number of offers we can reject before terminating the framework? right now if none of the offers match constraints - it seems mesos will keep resending them indefinitely.

Rob-Johnson commented:
@keymone you can ask mesos not to send you the same offer again by providing ``filters`` to the ``declineOffers`` call. maybe use that + a timeout of a minute or something?

Rob-Johnson commented on paasta_tools/cli/cmds/remote_run.py:
```diff
@@ -149,7 +163,11 @@ def paasta_remote_run(args):
             cmd_parts.extend(['--%s' % arg_key, value])
 
-    paasta_print('Running on master: %s' % cmd_parts)
+    constraints = map(lambda x: x.split(',', 2), args_vars['cons'])
```
sorry. ``[x.split(',',2) for x in args_vars['cons']]``

Rob-Johnson commented on paasta_tools/frameworks/native_scheduler.py:
```diff
@@ -125,22 +140,35 @@ def launch_tasks_for_offers(self, driver, offers):
-                launched_tasks.extend(tasks)
-            else:
-                driver.declineOffer(offer.id)
+            self.constraint_state_lock.acquire()
```
im fairly sure you can use the lock as a context manager - you can forget about explicitly dropping the lock that way.

```python
with self.constraint_state_lock.acquire():
     ...
```

keymone replied:
nice, thanks

Rob-Johnson commented on paasta_tools/frameworks/native_scheduler.py:
```diff
@@ -125,22 +140,35 @@ def launch_tasks_for_offers(self, driver, offers):
+                else:
+                    driver.declineOffer(offer.id)
+            except ConstraintFailAllTasksError:
+                paasta_print("Offer failed constraints for every task, rejecting 60s")
```
:+1:

Rob-Johnson commented on paasta_tools/frameworks/native_scheduler.py:
```diff
@@ -78,7 +88,7 @@ def __init__(self, service_name, instance_name, cluster,
                  service_config=None, reconcile_backoff=30,
                  instance_type='paasta_native', service_config_overrides={},
-                 reconcile_start_time=float('inf')):
+                 reconcile_start_time=float('inf'), constraints=[]):
```
mutable default args :( 

```
>>> class X:
...    def __init__(self, foo=[]):
...        self.foo = foo
...        print foo
... 
>>> x1 = X()
[]
>>> x1.foo.append('y')
>>> x1.foo
['y']
>>> x2 = X()
['y']
>>> x2.foo.append('foo')
>>> x2.foo
['y', 'foo']
```

you'll have to set the default arg as ``None`` and do ``if not constraints: self.constraints = []``

keymone replied:
crap, i just had a conversation about this couple days ago :) ruby habits will be dying slowly. thanks!

EvanKrall commented on paasta_tools/frameworks/native_scheduler.py:
```diff
@@ -240,7 +278,15 @@ def tasks_for_offer(self, driver, offer):
+            raise ConstraintFailAllTasksError
+
+        if len(tasks) > 0:
+            update_constraint_state(offer, self.constraints,
```
Won't we need to update our constraint state (or a temporary copy) every time we add a task to `tasks`? As written, I think this method could end up creating multiple tasks on the same offer and violating a UNIQUE constraint.

EvanKrall commented on paasta_tools/frameworks/native_scheduler.py:
```diff
@@ -125,22 +139,34 @@ def launch_tasks_for_offers(self, driver, offers):
-                launched_tasks.extend(tasks)
-            else:
-                driver.declineOffer(offer.id)
+            with self.constraint_state_lock:
```
Why do we need a lock here? I thought we only had one thread per scheduler.

keymone replied:
i don't know threading properties of mesos driver, if we're 100% sure no multiple statusUpdate/resourceOffer calls can be invoked in parallel we can drop the lock, for now it's a case of "better safe than sorry". let me know if you're certain we can remove the lock.

EvanKrall commented on paasta_tools/frameworks/native_scheduler.py:
```diff
@@ -400,6 +450,10 @@ def recreate_drain_method(self):
 
+    def reload_constraints(self):
+        if len(self.constraints) == 0 and self.service_config.get_constraints():
+            self.constraints = self.service_config.get_constraints()
```
Why copy the config's constraints into self?

Also, won't the `len(self.constraints) == 0` conditional prevent us from ever changing constraints?

EvanKrall commented on paasta_tools/paasta_remote_run.py:
```diff
@@ -87,7 +99,8 @@ def main(argv):
         dry_run=dry_run,
-        service_config_overrides={'cmd': command}
+        service_config_overrides={'cmd': command},
+        constraints=constraints
```
can we pass constraints in as part of `service_config_overrides`?

keymone commented on paasta_tools/cli/cmds/remote_run.py:
```diff
@@ -103,6 +108,15 @@ def add_subparser(subparsers):
         default=False,
     )
+    list_parser.add_argument(
+        '-X', '--cons',
```
cool, didn't know about that

keymone commented on paasta_tools/frameworks/constraints.py:
```diff
@@ -0,0 +1,84 @@
+            if offer_attr is None:
+                paasta_print("Attribute not found for a constraint: %s" % attr)
+                return False
+            elif not(CONS_OPS[op]([val, offer_attr.text.value,
```
argh, /me dumb. i thought lambda only supports single argument, probably because it only supports single statement. thanks

keymone commented:
@EvanKrall the problem with using tasks_with_flags for constraint state is that we don't modify it until we actually launch tasks, but we need to be updating constraint state for each task that fits in an offer. i felt like this would get hairy very quickly and decided to keep the state separate and copy+mutate+swap for each new offer if it results in launched tasks.

also we i've added offer reference to each task in tasks_with_flags to be able to update constraint state when task goes away.

thanks for review, i think i've addressed all the points but the locking one. i'm totally fine with dropping locks if we're sure it won't bite us later on.

keymone commented:
@EvanKrall @Rob-Johnson is this ready for shiping?

Rob-Johnson commented on paasta_tools/frameworks/constraints.py:
```diff
@@ -0,0 +1,80 @@
+from paasta_tools.utils import paasta_print
+
+
+def max_per(cv, ov, at, st):
```
can you write the names of these args out in full to make it a bit clearer?
